varlink/pid1 fixes for systemd-oomd (addresses #20330)

anitazha opened the pull request:
Attempt to fix the 2 issues brought up in #20330.

The first few commits address the issue of serializing the varlink sockets across systemctl daemon-reexec boundaries.
The last commit adds a check so that only systemd-oomd can use SubscribeManagedOOMCGroups.

DaanDeMeyer commented on src/shared/varlink.h:
```diff
@@ -158,6 +159,9 @@ unsigned varlink_server_current_connections(VarlinkServer *s);
 
 int varlink_server_set_description(VarlinkServer *s, const char *description);
 
+int varlink_server_serialize(VarlinkServer *s, FILE *f, FDSet *fds);
+int varlink_server_deserialize_one(VarlinkServer *s, const char *value, FDSet *fds);
```
In the interest of eventually making the varlink stuff public, I think we'd want this in a separate file so we can keep it private to systemd.

The other idea I had was adding varlink_server_get_socket_first() and varlink_server_get_socket_next() so that we can iterate over the sockets and then implement the serialization logic in pid1 on top of that.

keszybz commented on src/shared/varlink.c:
```diff
@@ -2299,10 +2324,16 @@ int varlink_server_listen_address(VarlinkServer *s, const char *address, mode_t
         if (r < 0)
                 return r;
 
+        addr_copy = strdup(address);
+        if (!addr_copy)
+                return log_oom_debug();
+
+        ss->address = TAKE_PTR(addr_copy);
```
`free_and_strdup(&ss->address, address)`

Hmm, so looking at this part, this block open-codes `varlink_server_listen_fd()`. Why not do:
```
r = varlink_server_listen_fd(s, fd);
if (r < 0)
   return r;

r = free_and_strdup(&s->sockets->address, address);
if (r < 0)
   return log_oom_debug();
```
I'm not saying that's better, just wondering.

anitazha replied:
I definitely considered it. But since `strdup()`/`free_and_strdup()` could fail, it felt like it went against the coding style if `varlink_server_listen_fd()` adds the `VarlinkServerSocket` object to the server's list if it failed to set the address.

keszybz commented on src/core/core-varlink.c:
```diff
@@ -476,30 +476,9 @@ static int manager_varlink_init_system(Manager *m) {
+        r = manager_setup_varlink_server(m, &s);
         if (r < 0)
-                return log_error_errno(r, "Failed to register varlink disconnect handler: %m");
+                return log_error_errno(r, "Failed to setup varlink server: %m");
```
"set up"

This would log twice, there's already logging going on in the function.

keszybz commented on src/core/core-varlink.c:
```diff
@@ -582,6 +561,42 @@ static int manager_varlink_init_user(Manager *m) {
+        assert(ret_s);
+
+        if (m->varlink_server)
+                return 1;
```
Hmm, what it the return convention, i.e. what is the difference between 0 and 1?

anitazha replied:
Hmmm I'm not sure. It looks like this all used to return 0 but was changed in 064a5c1438690f9454ad9bff05ec4ec2acfa67d2. @DaanDeMeyer is there a reason why we return both 0 and 1?

keszybz reviewed (commented):
Partial review, because I ran out of time. Will continue later.

keszybz commented on src/core/manager-serialize.c:
```diff
@@ -513,10 +519,21 @@ int manager_deserialize(Manager *m, FILE *f, FDSet *fds) {
-
                         if (strv_extend(&m->deserialized_subscribed, val) < 0)
                                 return -ENOMEM;
+                } else if ((val = startswith(l, "vl-ss-address="))) {
```
Hmm, if we already have a varlink server in `m->varlink_server`, shouldn't we skip deserialization here? In other words, I think we should not create a varlink server before deserialization, and here we should assert that `m->varlink_server` is not set, deserialize into `vl_server`, and then at the end move it over.

anitazha replied:
Yes, when testing this I was able to skip deserialization if `m->varlink_server` exists (for the `systemctl daemon-reload` case).

I noticed in the rest of the deserialization cases we close and replace the file descriptors regardless, so I did this for consistency. Thinking about it again, maybe the other code was written that way for code simplicity. If that's the case then yea I can update it to what you're suggesting.

keszybz commented on src/core/core-varlink.c:
```diff
@@ -202,11 +202,24 @@ static int vl_method_subscribe_managed_oom_cgroups(
 
         assert(link);
 
+        r = varlink_get_peer_uid(link, &uid);
```
Hmm, I thought we had a rule that PID1 was not allowed to do nss lookups because that is potentially blocking. But now I'm confused, because `chase_symlinks()` calls `uid_to_name()`, added in f1bda7ead20c210a28be5decad636e1d95d8c3ec. Does it need to be reverted? @poettering PTAL.

anitazha replied:
I guess that makes sense. But I see in this comment https://github.com/systemd/systemd/blob/c76691d708ac7fe13b7c4307c010d447fcbc3e9a/src/shared/specifier.c#L382-L389
that if UID is 0 it won't result in NSS. Which makes this okay?

keszybz replied:
I opened https://github.com/systemd/systemd/pull/24934 for the uid_to_name() uses in other places.

The use in `specifier_user_name()` is OK because we know that `uid==0` if we're running as PID1, and that is short-circuited to return "root". But here we're explicitly querying `uid`, so it can be anything and an nss lookup might occur.

keszybz replied:
Yeah, I'm pretty sure we can't do the lookup here.

anitazha replied:
> The use in `specifier_user_name()` is OK because we know that `uid==0` if we're running as PID1, and that is short-circuited to return "root". But here we're explicitly querying `uid`, so it can be anything and an nss lookup might occur.

Oh I see I didn't read that right. Ok I'll rework this commit and try something different.

anitazha commented:
Update according to review comments that seem unambiguous:
- downgrade logs to debug in `manager_setup_varlink_server()`
- skip deserialization if `m->varlink_server` exists
- use `free_and_strdup()`

anitazha commented:
Latest update changes and `assert` to `if` to properly handle serialization in user managers

keszybz commented on src/shared/varlink.c:
```diff
@@ -2263,11 +2274,24 @@ int varlink_server_listen_fd(VarlinkServer *s, int fd) {
+        return 0;
+}
+
+int varlink_server_listen_fd(VarlinkServer *s, int fd) {
```
Hmm, so shouldn't this take an optional `const char *address` to set the address? That way `varlink_server_listen_address()` would't need to duplicate the code.

anitazha replied:
Combine it into something like `int varlink_server_listen_object(VarlinkServer *s, int fd, const char *address, mode_t m)` ? This would mean either fd or address is passed. That would reduce the duplicate lines, though IMO it is more readable the way it is now.

yuwata commented on src/core/core-varlink.c:
```diff
@@ -476,30 +476,9 @@ static int manager_varlink_init_system(Manager *m) {
 
```
Commit message: pre-socket -> per-socket??

anitazha commented:
Changes:
- fix commit message typo
- use a command name check instead of username to avoid NSS lookup

bluca commented on src/core/core-varlink.c:
```diff
@@ -202,11 +203,23 @@ static int vl_method_subscribe_managed_oom_cgroups(
+        (void) get_process_comm(pid, &comm);
+
+        /* This is meant to be a deterrent and not actual security. */
+        if (!streq(comm, "systemd-oomd"))
```
Isn't it better to check which unit the process belongs to? Still not really a security measure, but a bit more robust given we don't allow duplicate units

poettering commented on src/shared/varlink.c:
```diff
@@ -2233,12 +2243,13 @@ static int connect_callback(sd_event_source *source, int fd, uint32_t revents, v
         int r;
 
         assert_return(s, -EINVAL);
         assert_return(fd >= 0, -EBADF);
```
internal functions should usually use assert(), not assert_return().

assert() → checks for programming errors in *our* code, and is strict and aborts

assert_return() → checks for programming errors in *foreign* code that calls ours, and is more lenient, and just debug logs and returns a error code.

The varlink code is of course currently only used inside of systemd itself, but the goal was to expose it as part of libsystemd.so to other projects, hence it uses assert_return() already.

Hence given this is now an internal function (static), please use simple assert() here, and then move the assert_return() asserts into the outer/non-static varlink_server_listen_fd() instead

poettering commented on src/core/manager-serialize.c:
```diff
@@ -537,5 +554,15 @@ int manager_deserialize(Manager *m, FILE *f, FDSet *fds) {
+                        return log_error_errno(r, "Failed to attach varlink connection to event loop: %m");
+
+                m->varlink_server = TAKE_PTR(vl_server);
+        }
```
hmm, what's the rationale for doing this late here? why not doing all this already in the loop above, if `m->varlink_server` is not set?

or in other words: whenever we see `vl-ss-address=`, then: 

1. allocate a varlink server object if there is none yet and add it to the event loop
2. add the deserialized fd to it

and that's it.

anitazha replied:
It was mostly done out of laziness of not wanting to refactor `varlink_server_attach_event()`. `varlink_server_attach_event()` sets up varlink server event loop and then loops over all of the sockets in its list to add those connect callbacks. I guess it makes more logical sense to add it as we see each socket address

poettering commented on src/shared/varlink.c:
```diff
@@ -2353,6 +2354,91 @@ void* varlink_server_get_userdata(VarlinkServer *s) {
+        LIST_FOREACH(sockets, ss, s->sockets) {
+                assert(ss->address);
+
+                fprintf(f, "vl-ss-address=%s", ss->address);
```
not a fan of needless abbreviation. Just call this `varlink-server-address"`?

poettering commented on src/core/core-varlink.c:
```diff
@@ -202,11 +203,23 @@ static int vl_method_subscribe_managed_oom_cgroups(
+
+        /* This is meant to be a deterrent and not actual security. */
+        if (!streq(comm, "systemd-oomd"))
+                return varlink_error(link, VARLINK_ERROR_PERMISSION_DENIED, NULL);
```
hmm, why not do actual security here? i.e. systemd-oomd already runs as its own user `systemd-oom`, so we could check the sender's UID and verify that it matches the UID of the `systemd-oom` user?

anitazha replied:
This was what I did in the first version of this PR. However @keszybz brought up that NSS lookups are blocking and that there is a policy against doing them pid1.

bluca replied:
checking that the pid is in the systemd-oomd service should be good enough?

poettering replied:
ah the nss issue is indeed a problem. Would be good to add a comment explaining why UID is not used there.

anitazha commented:
I think I addressed all of the latest review comments:
- updated from comm check to unit check
- refactor `varlink_server_attach_event()` so we can attach socket events as they are deserialized
- fix assert vs assert_return and naming nits

DaanDeMeyer commented:
I'd still put the serialize/deserialize functions in varlink-internal.h or so to clearly indicate these are not intended for public distribution if we ever make varlink.h public

anitazha commented on src/shared/varlink.c:
```diff
@@ -2574,3 +2575,94 @@ int varlink_server_set_description(VarlinkServer *s, const char *description) {
+        if (!address)
+                return log_oom_debug();
+
+        if (v[n] != ' ')
```
Yes. I'll update the serialize side to make it more clear that it will fail deserialization if this happens.

anitazha commented:
Addressed the latest round of comments from Yu and split out the varlink serialize/deserialize into `varlink-internal.h`

keszybz commented on src/core/core-varlink.c:
```diff
@@ -582,6 +567,36 @@ static int manager_varlink_init_user(Manager *m) {
         return 1;
 }
 
+int manager_setup_varlink_server(Manager *m, VarlinkServer **ret_s) {
```
This could be made `static`…

yuwata replied:
This is used in the serialization/deserialization code.

keszybz reviewed (commented):
LGTM.

yuwata commented on src/shared/varlink.c:
```diff
@@ -2380,6 +2380,30 @@ int varlink_server_shutdown(VarlinkServer *s) {
+
+        r = sd_event_add_io(s->event, &ss->event_source, ss->fd, EPOLLIN, connect_callback, ss);
+        if (r < 0)
+                goto fail;
```
It is not necessary to unref `ss->event_source`, as it is still NULL on failure.

yuwata commented on src/shared/varlink.c:
```diff
@@ -2380,6 +2380,30 @@ int varlink_server_shutdown(VarlinkServer *s) {
+        return 0;
+
+fail:
+        ss->event_source = sd_event_source_disable_unref(ss->event_source);
```
We usually use `_cleanup_` attribute in such case:
```c
_cleanup_(sd_event_source_unrefp) sd_event_source *s = NULL;

r = sd_event_add_io(s->event, &s, ss->fd, EPOLLIN, connect_callback, ss);
if (r < 0)
        return r;

r = sd_event_source_set_priority(s, ...);
if (r < 0)
        return r;

ss->event_source = TAKE_PTR(s);
return 0;
```

yuwata commented on src/core/core-varlink.c:
```diff
@@ -582,6 +567,36 @@ static int manager_varlink_init_user(Manager *m) {
         return 1;
 }
 
+int manager_setup_varlink_server(Manager *m, VarlinkServer **ret_s) {
```
Not a matter, but if a function has only one argument for taking return value, then we usually simply name it as `ret`. I think `_s` does not explain anything.

yuwata commented on src/core/manager-serialize.c:
```diff
@@ -516,7 +523,26 @@ int manager_deserialize(Manager *m, FILE *f, FDSet *fds) {
+                                r = varlink_server_attach_event(m->varlink_server, m->event, SD_EVENT_PRIORITY_NORMAL);
+                                if (r < 0) {
+                                        log_warning_errno(r, "Failed to attach varlink connection to event loop, ignoring: %m");
+                                        m->varlink_server = varlink_server_unref(m->varlink_server);
```
In this case, you can also use `_cleanup_` attribute and `TAKE_PTR()`.

yuwata commented on src/core/manager-serialize.c:
```diff
@@ -516,7 +523,26 @@ int manager_deserialize(Manager *m, FILE *f, FDSet *fds) {
+                                        continue;
+                                }
+
+                                deserialize_varlink_sockets = true;
```
Is this flag necessary??

anitazha replied:
I think so? It's only set if we have to create a new `m->varlink_server` during deserialization. This is to distinguish between cases when the manager deserialize code is called while `m->varlink_server` is still valid (such as during `systemctl daemon-reload`)

yuwata replied:
You mean, here we should only the first address and fd are deserialized?? If so, then maybe better to log something when we ignore the line, and please add a brief comment about that.

yuwata commented on src/shared/varlink.c:
```diff
@@ -2578,3 +2580,92 @@ int varlink_server_set_description(VarlinkServer *s, const char *description) {
+        if (v[n] != ' ')
+                return log_debug_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "Failed to deserialize VarlinkServerSocket: %s: %m", value);
+        p = v + n + 1;
+
+        v = startswith(p, "varlink-server-socket-fd=");
```
`v = startswith(v + n + 1, "...");`

yuwata commented on src/shared/varlink.c:
```diff
@@ -2578,3 +2580,92 @@ int varlink_server_set_description(VarlinkServer *s, const char *description) {
+        p = v + n + 1;
+
+        v = startswith(p, "varlink-server-socket-fd=");
+        if (v) {
```
I'd like to invert the condition here:
```
if (!v)
        return log_debug_errno(...);
```
then, we can drop `fd_found` flag.

yuwata commented on src/shared/varlink.h:
```diff
@@ -3,6 +3,7 @@
 
 #include "sd-event.h"
 
+#include "fdset.h"
```
Please drop this from the header, and move to `varlink-internal.h`

yuwata commented on src/shared/varlink-internal.h:
```diff
@@ -0,0 +1,5 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+#pragma once
+
```
Please include `varlink.h`, `fdset.h`, and `stdio.h`, otherwise, this header is not self consistent.

anitazha commented:
Addressed the latest round of comments from Yu and added a comment about why NSS is not used for the systemd-oomd unit check.

yuwata commented on src/shared/varlink-internal.h:
```diff
@@ -0,0 +1,9 @@
+#pragma once
+
+#include "fdset.h"
+#include "stdio.h"
```
`<studio.h>`

anitazha commented:
Fixed header and added comment.

keszybz commented on src/core/core-varlink.c:
```diff
@@ -203,10 +203,25 @@ static int vl_method_subscribe_managed_oom_cgroups(
+        if (!u)
+                return varlink_error(link, VARLINK_ERROR_PERMISSION_DENIED, NULL);
+
+        /* This is meant to be a deterrent and not actual security. The alternative is to check for the systemd-oom
```
I think that this comment is under-selling the protection provided here. Unit names are under control of root, and a unit name is not easy to fake. The attacker would need to move the PID into the systemd-oomd.service cgroup, which requires privileges again.
